### ISSUE TITLE
Data binding not updating listener variables on bind

### ISSUE THREAD

**Issue by chrisbanes** [airbnb/epoxy#502]
I wanted to open this issue since I don't really understand why the some of the data binding `setDataBindingVariables()` logic is implemented as:

``` java
if (((popupMenuListener == null) != (that.popupMenuListener == null))) {
  binding.setVariable(BR.popupMenuListener, popupMenuListener);
}
```

I'm guessing it is implicitly using `EpoxyAttribute(NoHash)` on my behalf.

Anyway my issue is that the listener is never updated on rebinds. Surely if a variable is set, it should be set on the binding. If the items are equals/==hashCode then sure don't update the binding, but the default so always be to set the binding.

At the moment it only updates the binding if the value goes from null to non-null and vice-versa, which is just confusing IMO.

**Comment by elihart** [comment]
Right, it's automatically applying [DoNotHash](https://github.com/airbnb/epoxy/blob/3aa26934eec27d12449c504ad3b58bcade0c0597/epoxy-annotations/src/main/java/com/airbnb/epoxy/EpoxyAttribute.java#L46)

I understand the confusion,  but imo the alternative is worse. If you have a bunch of models, and they all have anonymous click listeners, then every time you rebuild models the differ would think everything changed.

I think this expected behavior will differ depending on your use cases though. For me, generally we have immutable data and our click listener is never going to change behavior. If you do have mutable data this can be confusing and make life a bit harder.

This is also why I created the [OnModelClickListener](https://github.com/airbnb/epoxy/wiki/Epoxy-Models#click-listeners), so even though the click listener isn't rebound it always has access to the most recently created model and can get any data it needs off of it (similar to https://github.com/airbnb/epoxy/issues/499)

Databinding is the one place you don't have control over this setting though, and I would like it to be a choice. I think we could pretty easily modify the databinding annotations for providing layouts to add a boolean setting to disable this DoNotHash behavior by default

**Comment by chrisbanes** [comment]
Interesting. Your mention of _anonymous_ click listeners has made me think about another possible solution.

I could create a non-anonymous listener class which has a mutable property. I can then set an `onBind` listener on the model which update the listener instance.

**Comment by elihart** [comment]
Sure, that would work I think. That's basically what the `OnModelClickListener` is already doing

**Comment by chrisbanes** [comment]
Cool, this isn't an `OnClickListener` so I can't use that :)

**Comment by elihart** [comment]
Maybe it would be possible for Epoxy to provide a wrapper for any callback type object that would automatically provide the latest bound model.

**Comment by chrisbanes** [comment]
I couldn't get this to work by just updating the listener (couldn't work out why). In the end, the only way I could get this to work is by manually updating the Binding. I don't like it but it works.

https://github.com/chrisbanes/tivi/blob/master/app/src/main/java/app/tivi/showdetails/details/ShowDetailsEpoxyController.kt#L188

**Comment by shakil807g** [comment]
how can i use OnModelClickListener on Databinding layout ?? i got this error


`<?xml version="1.0" encoding="utf-8"?>
<layout xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto"
        xmlns:tools="http://schemas.android.com/tools">

    <data>

        <variable
                name="onModel"
                type="com.airbnb.epoxy.OnModelClickListener"/>


    </data>

    <android.support.constraint.ConstraintLayout
            android:layout_width="match_parent"
            android:layout_height="match_parent">

        <TextView
                android:text="TextView"
                android:layout_width="wrap_content"
                android:layout_height="wrap_content"
                android:id="@+id/textView"
                android:onClick="@{onModel}"
                app:layout_constraintTop_toTopOf="parent"
                app:layout_constraintStart_toStartOf="parent"
                android:layout_marginTop="226dp"
                app:layout_constraintEnd_toEndOf="parent"
                app:layout_constraintBottom_toBottomOf="parent"
                android:layout_marginBottom="266dp"/>
    </android.support.constraint.ConstraintLayout>
</layout>` 

data binding error ****msg:Cannot find the setter for attribute 'android:onClick' with parameter type com.airbnb.epoxy.OnModelClickListener<T extends com.airbnb.epoxy.EpoxyModel<?>,V>

**Comment by elihart** [comment]
What do you think of my previous suggestion?
```
 I think we could pretty easily modify the databinding annotations for providing layouts to add a boolean setting to disable this DoNotHash behavior by default
```

**Comment by chrisbanes** [comment]
@elihart Make sense to me 👍

**Comment by chrisbanes** [comment]
It's not actually the click listener which is the issue, it's the `SeasonPopupMenuListener` since that decides what items in the menu to show. 

The data class suggestion should work (good idea), but unfortunately it doesn't. I think something more tricky is happening but I haven't had time to dig into it yet.
